boat_simulator/nodes/data_collection/data_collection_node.py

```diff
@@ -0,0 +1,222 @@
+                ("write_period_sec", rclpy.Parameter.Type.DOUBLE),
+            ],
+        ),
+        all_parameters = self._parameters
+        for name, parameter in all_parameters.items():
+            value_str = str(parameter.value)
+            self.get_logger().debug(f"Got parameter {name} with value {value_str}")
+
+    def __init_storage_files(self):
+        """Initializes storage files for data logging."""
+        self.get_logger().debug("Initializing storrage files...")
+
+        if self.use_json:
+            self.__data_to_write = {}
+            self.__json_index_counter = 0
+            self.__start_time = self.get_clock().now().nanoseconds
+            json_file_path = "./" + self.file_name + ".json"
+            if os.path.exists(json_file_path):
+                self.get_logger().warn(
+                    f"JSON file with name {self.file_name} already exists. Overriding old file..."
+                )
+                os.remove(json_file_path)
+            self.__json_file = open(json_file_path, "a")
+            self.__json_file.write("[\n")  # Open JSON array
+
+        if self.use_bag:
+            self.__writer = rosbag2_py.SequentialWriter()
+            storage_options = rosbag2_py._storage.StorageOptions(
+                uri=self.file_name, storage_id="sqlite3"
+            )
+            converter_options = rosbag2_py._storage.ConverterOptions("", "")
+            self.__writer.open(storage_options, converter_options)
+
+    def __init_msg_types_dict(self):
+        """Prepare dictionary of all msg types with key name and value class"""
+        self.__msg_types_dict = {}
+        for name, cls in inspect.getmembers(custom_interfaces.msg, inspect.isclass):
+            if not name.startswith("_"):
+                self.__msg_types_dict[name] = cls
+
+    def __init_subscriptions(self):
+        """Initialize the subscriptions for this node. These subscriptions pertain to the topics
+        from which data will be collected."""
+        self.get_logger().debug("Initializing subscriptions...")
+
+        def create_subscription_with_callback(topic_name, msg_type):
+            return self.create_subscription(
+                msg_type=msg_type,
+                topic=topic_name,
+                callback=lambda msg: self.__general_sub_callback(msg, topic_name),
+                qos_profile=self.get_parameter("qos_depth").get_parameter_value().integer_value,
+            )
+
+        self.__record_topics_sub = []
+
+        # Assuming topics are specified as ['topic_name', 'topic_type']
+        for i in range(1, len(self.topics_to_record), 2):
+            topic_name = self.topics_to_record[i - 1]
+            msg_type_name = self.topics_to_record[i]
+
+            if msg_type_name not in self.__msg_types_dict:
+                self.get_logger().error(
+                    f"msg type {msg_type_name} does not exist. Please adjust the topics array in \
+                        globals.yaml"
+                )
```

Comment:
Should this crash the node or not? Also, should these checks be done before the for loop executes, or inside?